gapis/resolve/dependencygraph2/dce.go

```diff
@@ -0,0 +1,380 @@
+type DCEBuilder struct {
+	graph            DependencyGraph
+	requests         []api.SubCmdIdx
+	isLive           []bool
+	liveCmds         []api.Cmd
+	origCmdIDs       []api.CmdID
+	liveCmdIDs       map[api.CmdID]api.CmdID
+	numLiveInitCmds  int
+	orphanObs        []ObsNode
+	numDead, numLive int
+	deadMem, liveMem uint64
+}
+
+// NewDCEBuilder creates a new DCEBuiler using the specified dependency graph
+func NewDCEBuilder(graph DependencyGraph) *DCEBuilder {
+	return &DCEBuilder{
+		graph:      graph,
+		isLive:     make([]bool, graph.NumNodes()),
+		liveCmds:   make([]api.Cmd, 0, len(graph.Capture().Commands)),
+		origCmdIDs: make([]api.CmdID, 0, len(graph.Capture().Commands)),
+		liveCmdIDs: make(map[api.CmdID]api.CmdID),
+	}
+}
+
+// LiveCmdID maps CmdIDs from the original capture to CmdIDs in within the live commands.
+// If the old CmdID refers to a dead command, the returned command will refer to the next live command; if there is no next live command, api.CmdNoID is returned.
+func (b *DCEBuilder) LiveCmdID(oldCmdID api.CmdID) api.CmdID {
+	liveCmdID, ok := b.liveCmdIDs[oldCmdID]
+	if !ok {
+		return api.CmdNoID
+	}
+	return liveCmdID
+}
+
+// OriginalCmdIDs a live CmdID to the CmdID of the corresponding command in the original capture
+func (b *DCEBuilder) OriginalCmdID(liveCmdID api.CmdID) api.CmdID {
+	if liveCmdID.IsReal() {
+		liveCmdID += api.CmdID(b.numLiveInitCmds)
+	} else {
+		liveCmdID ^= initCmdBit
+	}
+	return b.origCmdIDs[liveCmdID]
+}
+
+// NumLiveInitiCmds returns the number of live commands which are initial commands.
+// (Initial commands are generated commands to recreate the initial state).
+func (b *DCEBuilder) NumLiveInitCmds() int {
+	return b.numLiveInitCmds
+}
+
+// LiveCmds returns the live commands
+func (b *DCEBuilder) LiveCmds() []api.Cmd {
+	return b.liveCmds
+}
+
+// Build runs the dead-code-elimination.
+// The subcommands specified in cmds are marked alive, along with their transitive dependencies.
+func (b *DCEBuilder) Build(ctx context.Context) error {
+	ctx = log.Enter(ctx, "DCEBuilder.Build")
+	numLiveCmds := 0
+	for _, cmd := range b.graph.Capture().Commands {
+		if cmd.Alive() {
+			numLiveCmds++
+		}
+	}
+	nodes := make([]NodeID, 0, len(b.requests)+numLiveCmds)
+	for i, cmd := range b.graph.Capture().Commands {
+		if cmd.Alive() {
+			nodeID := b.graph.GetNodeID(CmdNode{api.SubCmdIdx{uint64(i)}})
+			if nodeID != NodeNoID && !b.isLive[nodeID] {
+				b.isLive[nodeID] = true
+				nodes = append(nodes, nodeID)
+			}
+		}
+	}
+	for _, idx := range b.requests {
+		nodeID := b.graph.GetNodeID(CmdNode{idx})
+		if nodeID == NodeNoID {
+			return fmt.Errorf("Requested dependencies of cmd not in graph: %v", idx)
+		}
+		if !b.isLive[nodeID] {
+			b.isLive[nodeID] = true
+			nodes = append(nodes, nodeID)
+		}
+	}
+	t0 := dCE2Counter.Start()
+	b.markDependencies(nodes)
+	dCE2Counter.Stop(t0)
+
+	b.graph.ForeachNode(func(nodeID NodeID, node Node) error {
```

Comment:
I think we need to document the traversing order for `ForeachNode` and how `addCmoNodes` makes sure the order requirement is satisfied, and how the use of `orphanObs` interact with the order. This is an implicit state machine.

Reply:
I will add documentation to clarify this order, and the algorithm used by DCE more broadly.